core/src/main/java/ai/timefold/solver/core/impl/bavet/common/AbstractFlattenLastNode.java

```diff
@@ -41,117 +46,73 @@ public final void insert(InTuple_ tuple) {
             if (size == 0) {
                 return;
             }
-            List<OutTuple_> outTupleList = new ArrayList<>(size);
+            Map<Object, FlattenLastMapEntry<FlattenedItem_, OutTuple_>> outMap = CollectionUtils.newLinkedHashMap(size);
             for (FlattenedItem_ item : collection) {
-                addTuple(tuple, item, outTupleList);
+                addTuple(tuple, item, outMap);
             }
-            tuple.setStore(flattenLastStoreIndex, outTupleList);
+            tuple.setStore(flattenLastStoreIndex, outMap);
         } else {
             Iterator<FlattenedItem_> iterator = iterable.iterator();
             if (!iterator.hasNext()) {
                 return;
             }
-            List<OutTuple_> outTupleList = new ArrayList<>();
+            Map<Object, FlattenLastMapEntry<FlattenedItem_, OutTuple_>> outMap = new LinkedHashMap<>();
             while (iterator.hasNext()) {
-                addTuple(tuple, iterator.next(), outTupleList);
+                addTuple(tuple, iterator.next(), outMap);
             }
-            tuple.setStore(flattenLastStoreIndex, outTupleList);
+            tuple.setStore(flattenLastStoreIndex, outMap);
         }
     }
 
-    private void addTuple(InTuple_ originalTuple, FlattenedItem_ item, List<OutTuple_> outTupleList) {
-        OutTuple_ tuple = createTuple(originalTuple, item);
-        outTupleList.add(tuple);
-        propagationQueue.insert(tuple);
+    private void addTuple(InTuple_ originalTuple, FlattenedItem_ item,
+            Map<Object, FlattenLastMapEntry<FlattenedItem_, OutTuple_>> outTupleMap) {
+        var outTupleEntry = outTupleMap.computeIfAbsent(item, k -> new FlattenLastMapEntry<>(item));
+        outTupleEntry.add(() -> createTuple(originalTuple, outTupleEntry.value),
+                propagationQueue::insert,
+                propagationQueue::update);
     }
 
     protected abstract OutTuple_ createTuple(InTuple_ originalTuple, FlattenedItem_ item);
 
     @Override
     public final void update(InTuple_ tuple) {
-        List<OutTuple_> outTupleList = tuple.getStore(flattenLastStoreIndex);
-        if (outTupleList == null) {
+        Map<Object, FlattenLastMapEntry<FlattenedItem_, OutTuple_>> outTupleMap = tuple.getStore(flattenLastStoreIndex);
```

Comment:
The full type of this map is crazy enough that I'd create a new interface for it as a shorthand.